### DISCUSSION TITLE
Cursor blink setting overwritten by shell integration

### DISCUSSION THREAD

**Discussion by zoriya** [ghostty-org/ghostty#2812] [Q&A] [answered]
Setting `cursor-style-blink` to false has virtually no effect because shell integration will automatically turn on blinking. 

Here is the relevant snippet from zsh's shell integration.
```sh
      _ghostty_zle_line_init _ghostty_zle_line_finish _ghostty_zle_keymap_select() {
      case ${KEYMAP-} in
          # Blinking block cursor.
          vicmd|visual) builtin print -nu "$_ghostty_fd" '\e[1 q';;
          # Blinking bar cursor.
          *)            builtin print -nu "$_ghostty_fd" '\e[5 q';;
      esac
      }
```

I think the shell integration should:
- check for the blink config and not set blink if disabled, or
- use DEC mode 12 to set blink speratly from the cursor shape since DEC12 is ignored if `cursor-style-blink` is set.

PS: When suspending a tui (vim in my case) using ctrl-z, the cursor is not being changed by the shell integration idk why. Should I open another issue for this?

**Comment by mitchellh** [comment]
I'm not sure about this, so I converted it to a discussion to solicit more feedback.

Note that all of this can be disabled if you disable shell integration cursor handling with `shell-integration-features = no-cursor`.

**Reply by zoriya** [reply]
Yes but that disable the bar integration which i find pretty cool. Also since shell integration is enabled by default, changing `cursor-style-blink` does nothing so that's a bit weird.

I first created a post on discord to talk about this [here](https://discord.com/channels/1005603569187160125/1309672157533700188) before creating this issue.

**Reply by mitchellh** [reply]
> Also since shell integration is enabled by default, changing cursor-style-blink does nothing so that's a bit weird.

The easy solution I'd accept today would be to update the docs to note this similar to how we note that behavior already for `cursor-style`.

The separation is exactly what I want more feedback on before committing anything.

**Reply by talvdav** [reply]
If shell integration is on by default and shell-integration-features = no-cursor is not set by default then any option that will not work should be marked so in the documentation and/or should trigger a warning on config reload.

But i also think that any option set in the config should be respected regardless of the shell integration status.

**Reply by olejorgenb** [reply]
I don't understand how this is not a clear (minor) bug?

If I set `cursor-style-blink = false`  surely that should affect the cursor style for this integration as well? I could understand it being less clear if the mode indication was using the blinking state (insert mode: blinking, normal mode: non-blinking), but it's not.

**Comment by samuelmasuy** [comment]
At least with Kitty and shell integration, when setting `cursor_blink_interval 0`, the behaviour is to `check for the blink config and not set blink if disabled`. I would personally like this behaviour, because I still want the shell integration for the cursor feature. Indeed, the cursor feature also changes the text cursor to a bar when editing shell command line.
On that note, the doc for `shell-integration-features` is a bit misleading since it does not talk about the feature I just referred to:
```
# Available features:
# 
#   * `cursor` - Set the cursor to a blinking bar at the prompt.
```

**Reply by cwegener** [reply]
I totally agree.

e.g. in my vimrc, my cursor is set to change to a bar when when in insert mode.
it would be awesome if ghostty did the same. right now in ghostty I have to chose between one of the two:
1. have the cursor on the zsh prompt change to a BLINKING bar when I am in insert mode AND have a BLINKING block cursor when in normal mode
2. disable a the cursor shell integration in ghostty altogether and always have a block cursor on the zsh prompt everywhere regardless of mode

at the moment, I have to chose option 2 until blinking is fixed ..

**Reply by magnusriga** [reply]
@cwegener Did you find a solution? I am facing the same exact issue.

**Reply by cwegener** [reply]
@magnusriga I am happily living with option 2 and have disabled cursor shell integration in zsh. my old terminal emulators never had cursor shell integration, so it's not really a feature that I'm missing.

**Reply by magnusriga** [reply]
Got it @cwegener . How are you able to tell if you are in Insert mode or Normal mode, if you always have a block cursor?

**Reply by cwegener** [reply]
I'm not actually using vi-mode in zsh. I'm sticking to emacs-mode.

The shell cursor integration feature in ghostty did make me investigate vim-mode again and that's where my original comment came from.

But in the end, emacs-mode in zsh is just what I'm used to.

**Comment by nornagon** [comment]
Definitely very annoying not to have an easy way to turn off cursor blinking. It's counterintuitive that setting cursor-style-blink = false in the config seems to do nothing.

**Reply by denolfe** [reply]
I just ran into this and found it rather unintuitive as well. I would expect explicitly setting blink to false to be respected over anything else.

**Reply by rtfeldman** [reply]
I don't have a preference about how the setting should work, but I find blinking cursors really distracting (I seem to be abnormally easily distracted by motion compared to most people, at least based on talking to them; I also turn off all animated gifs in chat apps) and so the setting I was looking for was "never blink the cursor under any circumstances."

I actually didn't know zsh had any say in that!

I used `shell-integration-features = no-cursor` from [this post below](https://github.com/ghostty-org/ghostty/discussions/2812#discussioncomment-11688463) (thanks!) but was disappointed to see that now I get a block cursor instead of the nice thin bar. It doesn't blink, so I'll keep using it that way, but I'd love a non-blinking bar if there's some way to get it!

**Reply by thiagowfx** [reply]
Richard: set `cursor-style = bar` as well

**Reply by rtfeldman** [reply]
Beautiful, that brought it back - thank you! 😃

**Reply by b11e** [reply]
I find that the `shell-integration-features = no-cursor` setting doesn't stick.
Does it go back to blinking for you guys after some time?

**Comment by thiagowfx** [comment]
I use [grml-zsh-config](https://grml.org/zsh). When opening Ghostty, it starts with a bar+blinking cursor(1).

When spawning a new `zsh` shell within `zsh`, it changes to a block+non-blocking cursor(2).

(2) is the desired behavior – iTerm2 and Terminal.app both do it out-of-the-box. I am not sure why Ghostty sets a bar+blinking cursor, whilst the other terminal emulators do not.

This discussion / issue seems to track down the root cause. I would suggest converting it back to an Issue. Happy to provide logs and/or more info if it helps!

**Reply by thiagowfx** [reply]
Workaround to achieve (2):

```
cursor-style-blink = false
shell-integration-features = no-cursor
```

Source: https://github.com/ghostty-org/ghostty/discussions/3836

**Comment by cbrunnkvist** [comment]
Oh man, I was **THIS** close to opening another Issue for "_Config file isn't working at all on macOS_" 
because the _first_ thing I decided needed changing upon first launch was the incessant blinking and indistinct "|" style cursor(_1_):

```
cursor-style-blink = false
cursor-style=block
```

Anyway, it turns out the issue was the shell integration (whatever it means) and the workaround suggested above, worked.

I also would like to note that the "Ghostty > Reload configuration" menu option doesn't reset the currently open terminal an/or reapply the shell-integration-features setting. This is another reason I thought it was completely ignoring the config file - there's no logging or anything, not even if I make typos in the config.

(_1): I've never seen that in an interactive prompt where the cursor is nominally at the end of data - in a word processor, yes okay, because the nature of text editing is completely different there_

**Comment by jessejanderson** [comment]
Just chiming in to say I also thought this was broken since it was the first thing I noticed and tried to configure.

While admittedly a very minor thing (and simple enough to fix), it was a pretty sour start to trying out something new. Call me fickle, but had I not discovered this thread I might have just been annoyed enough to remove the app and switch back to my old terminal app without giving ghostty a fair shake. 🙃

Definitely something that should—at the very least—be mentioned in the docs next to the `cursor-style` and/or `cursor-style-blink` info.

**Comment by b11e** [comment]
Edit: ok it only blinks inside tmux now.
Edit2: i forgot to kill tmux-server and then restart 🤦‍♂️



I don't know what's going but it still blinks even with
`shell-integration-features = no-cursor`
it works only temporarily when it does (i have to quit and restart) then I come back to ghostty later and it's blinking again.
Most times I reload and restart, it just sits there blinking and it's driving me up a wall.


running tmux + nvim + ghossty
config:
```font-family = DankMono Nerd Font
font-thicken = true
font-size = 26
cursor-style = block
cursor-style-blink = false
cursor-invert-fg-bg = true
shell-integration-features = no-cursor
```

**Reply by risico** [reply]
Same here, although not sure if it is a similar issue, my cursor starts to blink after I enter nvim from within tmux, and then it keeps blinking even after I quit nvim. If I go into nvim without tmux then no blink, so I am not sure how nvim triggers the blink just in tmux. 

But yeah, +1 on having `cursor-style-blink` just say hard no to any blinking anywhere.

**Reply by 00-kat** [reply]
Neovim doesn't seem to reset the cursor blink state when you exit, and Tmux doesn't have shell integration.  At least, that's my guess as to what this is.

**Reply by risico** [reply]
I am not sure what config in nvim deals with that? My cursor is set to be block in all modes even.

**Reply by 00-kat** [reply]
`guicursor` I think.

**Comment by gaoxiaoliangz** [comment]
Any update on this? I want to have `shell-integration-features = cursor` to show the zsh vi mode status, but also keep the cursor from blinking. I set `cursor-style-blink = false` but the cursor still blinks.

**Reply by chris4prez** [reply]
This is how I would like it to also work too. I want shell integration for cursor but want to disable blink which I find very distracting

**Reply by barrettruth** [reply]
Same! Anyone know how we can make this happen? The settings below basically imply that a cursor shell integration *must* mean blinking - i.e. that there are no other special cursor behaviors a shell may have.

So not only should this issue be resolved (I feel that we have a non-default behavior as default) the structure of the settings should also be refactored.

![image](https://github.com/user-attachments/assets/0313d3c4-59aa-4ce9-aba2-e2c9c248f3e2)

**Comment by thecsw** [comment]
If you're using fish, at the very least you can put `set fish_cursor_unknown block` in your `config.fish`.

**Reply by 00-kat** [reply]
Also: if you use Vi mode, a few other variables are:  `fish_cursor_insert` `fish_cursor_visual` `fish_cursor_default` `fish_cursor_replace` `fish_cursor_unknown` `fish_cursor_external` `fish_cursor_replace_one` `fish_cursor_selection_mode` (and these variables are also mentioned [over here]).

You may need one of these too (until [Fish v4 which adds Ghostty to the allowlist]):
```sh
if status is-interactive
    if string match -q -- '*ghostty*' $TERM
        set -g fish_vi_force_cursor 1
    end
end
```

[over here]: https://fishshell.com/docs/current/interactive.html#vi-mode-commands
[Fish v4 which adds Ghostty to the allowlist]: https://github.com/fish-shell/fish-shell/pull/10685

**Reply by wolverian** [reply]
> If you're using fish, at the very least you can put `set fish_cursor_unknown block` in your `config.fish`.

This only works in Vi mode, however.

**Comment by barrettruth** [comment]
Fully functional zsh solution (vi mode support), requiring no ghostty setting updates:

```zsh

function __set_beam_cursor {
    echo -ne '\e[6 q'
}

function __set_block_cursor {
    echo -ne '\e[2 q'
}

function zle-keymap-select {
  case $KEYMAP in
    vicmd) __set_block_cursor;;
    viins|main) __set_beam_cursor;;
  esac
}
zle -N zle-keymap-select

function zle-line-init {
    __set_beam_cursor
}
zle -N zle-line-init

precmd() {
    __set_beam_cursor  # doesn't have to be in precmd - can put outside a function if you like
}
```

**Reply by magnusriga** [reply]
+1, that worked. Thank you!

@barrett-ruth Is it necessary to set the cursor in both `precmd` and `zle-line-init`, would not one of them suffice? From the ZSH docs:

> [precmd](https://zsh.sourceforge.io/Doc/Release/Functions.html#Hook-Functions)
> Executed before each prompt. Note that precommand functions are not re-executed simply because the command line is redrawn, as happens, for example, when a notification about an exiting job is displayed. 

> [zle-line-init](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#Special-Widgets)
> Executed every time the line editor is started to read a new line of input.

Also, re. `precmd`, should we not append to the *hook function*, instead of re-defining `precmd` (since it might be used elsewhere), like this:

```bash
builtin typeset -ag precmd_functions
precmd_functions+=(__set_beam_cursor)
```

Edit:

I tested the below, and so far it seems to work:

```bash
# Prevent blinking cursor.
function __set_beam_cursor {
    echo -ne '\e[6 q'
}

function __set_block_cursor {
    echo -ne '\e[2 q'
}

function zle-keymap-select {
  case $KEYMAP in
    vicmd) __set_block_cursor;;
    viins|main) __set_beam_cursor;;
  esac
}
zle -N zle-keymap-select

precmd_functions+=(__set_beam_cursor)
```

I made sure `no-cursor` was set, to not run code similar to the above in the ghostty shell integration script as well.

Also manually set `export GHOSTTY_SHELL_INTEGRATION_NO_CURSOR=1` in `.zshrc`, to make the shell operate as expected in ssh (where Ghostty does not currently set that variable, and its siblings).

**Reply by olejorgenb** [reply]
Yeah, I already have something similar in my config and wasted 30min debugging why the `cursor-style-blink = false` config did not work.

IMO - this type of shell integration should not be on by default when it has these gotchas.

And is it really the terminal's business at all to provide such shell-specific features? I'm torn, because I do like when software have nice defaults (and using the cursor shape to indicate mode is a nice default). But these defaults are not really terminal defaults, they are shell defaults. So I feel like this should not be so "aggressively" pushed by Ghostty. Unless there are some tangible benefit of having Ghossty specific integration? My existing zsh config seem to work just fine though.

**Reply by barrettruth** [reply]
thanks a lot, just learned a goot bit about zsh @magnusriga .

**Comment by jparise** [comment]
I'd like to address this, and I implemented some options to see how they felt:

1. Simply make the shell integration respect `cursor-style-blink`. When the `cursor` shell feature is enabled, we'll still always use the bar (beam) cursor shape, but it'll be steady when `cursor-style-blink = false`. This would be a behavior change for those users (some want it, others might not). 
2. Extend `shell-integration-features` to allow `cursor=<value>`, to independently control the shell integration cursor. This would be the most flexible solution, but it's also more complex, less discoverable, and still effectively ignores `cursor-style-blink` (unless we also use `cursor-style-blink` as a default unless overridden by some form of `cursor=...`; you get the idea).

I have (1) fully implemented and tested, and I'm considering building out (2) because I personally like the independent control (although being the _maintainer_ also makes me prefer the simpler solution).

So, I thought this was a good point to pause and ask the community your thoughts before going further.

**Reply by taylrfnt** [reply]
I think Option 1 (respect the configuration) makes the most sense after reading through the docs and config options.

Agreed Option 2 has some merits, but I think it doesn't align with the UX philosophy of the project (ignoring, conflicting, or duplicating config - however you want to frame it).

Just my two cents!

**Reply by barrettruth** [reply]
#1

Thanks for response and thinking about something so "trivial" as cursor shape intently.

**Reply by jparise** [reply]
#10643 implements Option 1. I also have Option 2 working, but I think that's best left to a separate follow-up change (i.e. the first change fixes the core issue and the second (re)provides the ability to configure the shell's blink-ability).

**Reply by rhodes-b** [reply]
https://github.com/ghostty-org/ghostty/issues/8681

note a 3rd option which is already an issue that would wouldn't require any changes to the shell-integration as it would be handled natively inside of the configuration parsing

**Reply by jparise** [reply, marked as answer]
#10643 implements Option 1. That's in tip and will be in 1.3. I think that closes out this particular discussion.

#10776 extends the configuration support even further (Option 2). That's still pending review.
